### ISSUE TITLE
`Concatenate` implementation generates runtime exception if parameterized by `...`

### ISSUE THREAD

**Issue by erictraut** [python/typing_extensions#48]
The following code should work without a runtime exception.

```python
from typing import Any, Callable
from typing_extensions import ParamSpec, Concatenate


P = ParamSpec("P")
MyAlias = Callable[Concatenate[int, P], Any]

x: MyAlias[...]
```

It runs fine on Python 3.10 if using `typing.Concatenate` but fails on Python 3.9 if using `typing_extensions.Concatenate`. The runtime exception is:
```
TypeError: Parameters to generic types must be types. Got Ellipsis.
```

**Comment by Fidget-Spinner** [comment]
This seems like a bug in Python's typing module itself.
```python
# In 3.9
typing.Callable[T, int][...]

TypeError: Parameters to generic types must be types. Got Ellipsis.
```
The problem is that `Callable` uses the default `_GenericAlias`'s substitution behavior pre 3.10. Which means its checks are too restrictive. We had to intentionally loosen the checks in 3.10 to get around this.

I don't think we can fix this from `typing_extensions` unless we monkey patch `typing._GenericAlias.__getitem__`. I don't feel too good doing that though.

**Comment by JelleZijlstra** [comment]
This is now documented as a limitation at https://typing-extensions.readthedocs.io/en/latest/#Concatenate. If there is a reasonable way to fix it, I'd accept a PR.

**Comment by Daraan** [comment]
I am currently looking at this again and the statement that "It runs fine on Python 3.10 if using typing.Concatenate" confused be a bit.
In hindsight this should only hold for 3.10.0-3.10.2 afterwards it also does not run on the other 3.10 `typing` versions either.

---

This should be fixable for the rest of the 3.10 versions if `typing.Concatenate` is monkey patched to use the `typing_extensions.Concatenate` variant combined with `_concatenate_getitem` from #442 that accepts an `...` argument.

Whats then left to do is to use `typing_extensions._ConcatenateGenericAlias` for 3.10.3+ < 3.11 instead of `typing._ConcatenateGenericAlias`.
This will fix the usability aspect

This then will cause a regression in the `__args__` attribute as `typing_extensions._ConcatenateGenericAlias` inherits from list which will be unpacked instead of staying an instance in the typing module (iirc there is an iterable check).
From the comment this should only be necessary for Python 3.9.(<2?).

In short the regression will be these two tests:
```python
                # skip for 3.10.3+ < 3.11
                G2 = klass[int, Concatenate[int, P_2]]
                G3 = klass[int, Concatenate[int, ...]]
                self.assertEqual(G2.__args__, (int, Concatenate[int, P_2]))
                self.assertEqual(G3.__args__, (int, Concatenate[int, ...]))
```

I think if the `_ConcatenateGenericAlias` in 3.10 does not inherit from a list these tests should also pass, which sound doable.

**Comment by Daraan** [comment]
As a summary:

#479 covered this Issue only for 3.10.3+

For 3.8 & 3.9 I see no other way around monkey patching `typing._GenericAlias.__getitem__` to fix this issue.
The essential part of the monkey patch could be similar to this, maybe with more checks on the origin:

```diff
- params = tuple(_type_check(p, msg) for p in params)
+ params = tuple(typing._type_check(p, msg) if p is not ... else ...
+                       for p in params)
```
